occurrence-ws/src/main/java/org/gbif/occurrence/ws/config/ChecklistAwareSearchRequestHandlerMethodArgumentResolver.java

```diff
@@ -0,0 +1,228 @@
+          checklistParameters.add(new OccurrenceSearchParameter(rank + "_KEY", String.class));
+        });
+      } catch (Exception e) {
+        log.error("Failed to get metadata for checklist {}", checklistKey, e);
+      }
+    }
+  }
+
+  private List<OccurrenceSearchParameter> getChecklistParameters(OccurrenceSearchRequest request) {
+
+    List<OccurrenceSearchParameter> checklistParameters = new ArrayList<>();
+
+    // add support for dynamic facets for ranks ....
+    if (request.getParameters().containsKey(OccurrenceSearchParameter.CHECKLIST_KEY)){
+      // get a list of recognised ranks for this checklist
+      Set<String> checklistKeys = request.getParameters().get(OccurrenceSearchParameter.CHECKLIST_KEY);
+      if (!checklistKeys.isEmpty()) {
+        loadChecklistRankParams(checklistKeys.iterator().next(), checklistParameters);
+      }
+    }
+    return checklistParameters;
+  }
+
+  /**
+   * Iterates over the params map and adds to the search request the recognized parameters (i.e.: those that have a
+   * correspondent value in the P generic parameter).
+   * Empty (of all size) and null parameters are discarded.
+   */
+  @Override
+  protected void setSearchParams(OccurrenceSearchRequest searchRequest, Map<String, String[]> params) {
+
+    // look for a checklist param first....
+    List<OccurrenceSearchParameter> checklistParameters = getChecklistParameters(params);
+
+    for (Map.Entry<String, String[]> entry : params.entrySet()) {
+
+      String param = entry.getKey();
+
+      OccurrenceSearchParameter p = findSearchParam(param);
+
+      // look for dynamic rank facet names
+      p = getOccurrenceSearchParameter(checklistParameters, param, p);
+
+      if (p != null) {
+        final List<String> list =
+          entry.getValue() != null ? Arrays.asList(entry.getValue()) : Collections.emptyList();
+        for (String val : removeEmptyParameters(list)) {
+          // validate value for certain types
+          SearchTypeValidator.validate(p, val);
+          searchRequest.addParameter(p, val);
+        }
+      }
+    }
+  }
+
+  private OccurrenceSearchParameter getOccurrenceSearchParameter(List<OccurrenceSearchParameter> checklistParameters, String param, OccurrenceSearchParameter p) {
+    if (p == null) {
+      String normedType = param.toUpperCase().replaceAll("[. _-]", "");
+      // check if this is a checklist parameter
+      for (OccurrenceSearchParameter cp : checklistParameters) {
+        if (cp.name().replaceAll("[. _-]", "").equalsIgnoreCase(normedType)) {
+          p = cp;
+          break;
+        }
+      }
+    }
+
+    if (p == null) {
+      Optional<Integer> depthOpt = extractTaxonDepth(param);
+      if (depthOpt.isPresent()) {
+        p = new OccurrenceSearchParameter("TAXON_DEPTH_" + depthOpt.get(), String.class);
+      }
+    }
+    return p;
+  }
+
+  private static String getFirstIgnoringCase(String parameter, Map<String, String[]> params) {
```

Comment:
isn't this the same as the one that exists in one of the parents? I guess we can make it protected in the parent and use it here

Reply:
ive changed the visibility now, and removed this